fix(ivy): ensure static styling is properly inherited into child components

matsko opened the pull request:
Angular supports having a component extend off of a parent component.
When this happens, all annotation-level data is inherited including styles
and classes. Up until now, Ivy only paid attention to static styling
values on the parent component and not the child component. This patch
ensures that both the parent's component and child component's styling
data is merged and rendered accordingly.

Jira Issue: FW-1081

matsko commented:
https://test.corp.google.com/ui#id=OCL:236013027:BASE:236013064:1551312718590:4e547afd

matsko commented:
https://test.corp.google.com/ui#id=OCL:236013027:BASE:236134054:1551373984205:bfd2447e

kara reviewed (changes requested):
Let's chat more about this in the office tomorrow? Some preliminary thoughts below

kara commented on packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts:
```diff
@@ -346,7 +346,7 @@ describe('compiler compliance: bindings', () => {
           hostBindings: function HostAttributeDir_HostBindings(rf, ctx, elIndex) {
             if (rf & 1) {
-              $r3$.ɵelementHostAttrs(ctx, $c0$);
+              $r3$.ɵelementHostAttrs(HostAttributeDir, ctx, $c0$);
```
Also wouldn't the type be the same for all these invocations? Maybe we just pass in the directive index instead and we can look up the definition, which has the type on it.

matsko committed:
fix(ivy): ensure static styling is properly inherited into child components

Angular supports having a component extend off of a parent component.
When this happens, all annotation-level data is inherited including styles
and classes. Up until now, Ivy only paid attention to static styling
values on the parent component and not the child component. This patch
ensures that both the parent's component and child component's styling
data is merged and rendered accordingly.

Jira Issue: FW-1081

matsko commented:
https://test.corp.google.com/ui#id=OCL:236013027:BASE:236378090:1551479075634:3003d0fa

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -3372,3 +3383,13 @@ function handleError(lView: LView, error: any): void {
 }
+
+function attrsStylingIndexOf(attrs: TAttributes, startIndex: number): number {
+  for (let i = startIndex || 0; i < attrs.length; i++) {
```
When would `startIndex` be falsy here? The type seems to indicate it will always be a number.

kara commented on packages/core/src/render3/instructions.ts:
```diff
@@ -3372,3 +3383,13 @@ function handleError(lView: LView, error: any): void {
   errorHandler && errorHandler.handleError(error);
 }
+
+function attrsStylingIndexOf(attrs: TAttributes, startIndex: number): number {
```
Seems like this could probably live in a util somewhere (`instructions.ts` is getting pretty heavy)

kara commented on packages/core/src/render3/styling/util.ts:
```diff
@@ -41,9 +41,13 @@ export function createEmptyStylingContext(
-  context[StylingIndex.DirectiveRegistryPosition].push(directiveRef, -1, false, null);
+  const dirs = context[StylingIndex.DirectiveRegistryPosition];
+  const index = dirs.length;
+  dirs.push(directiveRef, -1, false, null);
```
Nit: It's not clear why you're pushing in `-1, false, null`. Add a comment for what these are?

kara reviewed (commented):
Since this change was intended to fix some Material tests, can you update the blocklist to turn those tests on?  Affected tests should be labeled FW-1081:

https://github.com/angular/angular/blob/master/tools/material-ci/angular_material_test_blocklist.js#L534

kara reviewed (approved):
LGTM

kara commented:
@matsko FYI, rebased this to fix the blocklist

kara commented:
[presubmit](http://test/OCL:236397238:BASE:236397288:1551485683291:5b315f4d)

kara commented:
merge-assistance: compiler changes are no longer part of this PR

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -620,15 +621,21 @@ export function elementStart(
-      tNode.stylingTemplate = initializeStaticStylingContext(attrs);
+    // to be patched and constructed at various points, but only up until the styling
+    // template is first allocated (which happens when the very first style/class binding
+    // value is evaluated). When the template is allocated (when turns it into a context)
```
```suggestion
    // value is evaluated). When the template is allocated (when it turns into a context)
```

mhevery commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -433,11 +433,12 @@ export interface InitialStylingValues extends Array<string|boolean|null> {
  */
 export const enum InitialStylingValuesIndex {
   DefaultNullValuePosition = 0,
```
could you add text which explain what each of the enums are used for.

mhevery commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -110,29 +98,39 @@ export function patchContextWithStaticAttrs(
+ *  2) if the value was set already by the template, component or directive, but the
+ *     new value is set on a higher level (i.e. a sub component which extends a parent
+ *     component sets its value after the parent has already set the same one)
+ *  3) if the same directive provides a new set of styling values to set
```
add `@props` docs

mhevery commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1956,3 +1956,14 @@ function registerMultiMapEntry(
   cachedValues.push(0, startPosition, null, count);
 }
+
+function registerStaticStyle(
```
docs please

mhevery commented on packages/core/src/render3/styling/util.ts:
```diff
@@ -37,13 +37,45 @@ export function createEmptyStylingContext(
+ * For each and every `[style]`, `[style.prop]`, `[class]`, `[class.name]` binding
+ * (as well as static style and class attributes) a directive, component or template
+ * is marked as the owner. When an owner is deteremined (during first template pass
+ * creation block) the owner is allocated into the styling context so any that once
```
awkward sentence structure. `into the styling context so any that once`

mhevery commented on packages/core/src/render3/styling/util.ts:
```diff
@@ -37,13 +37,45 @@ export function createEmptyStylingContext(
+  const dirs = context[StylingIndex.DirectiveRegistryPosition];
+  const i = dirs.length;
+
+  // we pre-emptively make space into the directives array and then
```
```suggestion
  // we preemptively make space into the directives array and then
```

mhevery reviewed (approved):
minor cleanup

JoostK commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1957,13 +1962,29 @@ function registerMultiMapEntry(
+    index = staticStyles.length;
+    staticStyles.push(null, null, null);
+    staticStyles[index + InitialStylingValuesIndex.PropOffset] = prop;
+  }
```
This closing bracket looks out-of-place.

JoostK commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1957,13 +1962,29 @@ function registerMultiMapEntry(
-  values[i + InitialStylingValuesIndex.DirectiveOwnerOffset] = directiveOwnerIndex;
-  return i;
+/**
+ * Inserts or updates and existing entry in the provided `staticStyles` collection.
```
```suggestion
 * Inserts or updates an existing entry in the provided `staticStyles` collection.
```

JoostK commented on packages/core/src/render3/styling/util.ts:
```diff
@@ -45,13 +45,16 @@ export function createEmptyStylingContext(
- * any style and class bindings are assigned then they are tied to that directive.
+ * is marked as the owner. When an owner is determined (this happens when the template
+ * is first passed over) the directive owner is allocated into the styling context. When
+ * this happens, each owner gets its own index value. This then ensures that any that
```
```suggestion
 * this happens, each owner gets its own index value. This then ensures that
```

matsko commented:
http://test/OCL:236013027:BASE:236668438:1551721051031:6e806984

JoostK commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -432,12 +432,34 @@ export interface InitialStylingValues extends Array<string|boolean|number|null>
+  /** The first value is always `null` so that `styles[0] == null` for unassigned values  */
   DefaultNullValuePosition = 0,
+
+  /** Used for non-styling code to examine what the style or className string is:
```
Insert newline at beginnen of JSDoc

JoostK commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -432,12 +432,34 @@ export interface InitialStylingValues extends Array<string|boolean|number|null>
+   * classes: ['foo', true, 0, 'bar', false, 0, 'baz', true, 0]
+   *    => initialClasses[CachedStringValuePosition] = 'foo bar';
+   *
+   * Note that this value is `null` be default and it will only be populated
```
```suggestion
   * Note that this value is `null` by default and it will only be populated
```

matsko commented:
Thank you @JoostK

matsko commented:
https://tap.corp.google.com/ui#id=OCL:236013027:BASE:236670850:1551721884802:be9cdb4f

matsko commented:
https://tap.corp.google.com/ui#id=OCL:236013027:BASE:236689574:1551727104991:a28e154e
